Fix race conds when running the OCaml test suite with parallel make

ZackerySpytz opened the pull request:
Don't try to rm files that have been checked out of the SWIG library.
Don't repeatedly attempt to compile swig.ml, swig.mli, and swigp4.ml.

wsfulton commented:
This won't properly fix the race conditions. The firsts N job when running make -jN will attempt to run the same operations on each of the commands in OCAMLCORE at nearly the same time. 

We don't have any other target languages that needs to build some sort of core component, so unfortunately there isn't a reference to use within SWIG. However, the correct solution would be to use Make dependencies, but may complicate the makefiles somewhat.

wsfulton commented:
I just remembered, I did solve this problem before. Use the generation of CommentParser.class from CommentParser.java in the Java test-suite as a reference. A number of tests (doxygen tests) have a dependency on CommentParser.class. See commit 4e27fdbfc123dc4a0d41f4b5b27cb3cf4a792206. It relies on multiple make rules to add additional dependencies for a target: https://www.gnu.org/software/make/manual/html_node/Multiple-Rules.html#Multiple-Rules.

ZackerySpytz commented:
Thank you, @wsfulton, for the comments. I've updated the PR based on the commit you've linked.

Currently, all of the OCaml runtime tests depend on a single `runme` executable, so that's another race condition that needs to be fixed. I've updated Examples/test-suite/ocaml/Makefile.in to generate unique executable names (based on what is done in Examples/test-suite/go/Makefile.in).

ZackerySpytz commented:
`make check-ocaml-examples` and `make check-ocaml-test-suite` are failing on Travis for unrelated reasons. It seems that the changes in 4e4bb54fe7f01432825a1e0bcc1ffde7bd822271 / PR #1393 are somehow not applied, even though I merged this branch (OCaml-make-j-race-conds) with master. I'm not sure why this is.

wsfulton commented on Examples/test-suite/ocaml/Makefile.in:
```diff
@@ -85,9 +86,24 @@ include $(srcdir)/../common.mk
+swigp4.cmi:
+	env SWIG_LIB=$(SWIG_LIB_DIR) $(SWIGEXE) -ocaml -co swigp4.ml
+	$(OCC) -I $(OCAMLP4WHERE) -pp "camlp4o pa_extend.cmo q_MLast.cmo" -c swigp4.ml
+extra_obs: swig.cmi swig.cmo swigp4.cmi
```
Can you use full English, so extra_objects (I think you mean obs==objects, and a .cmi file is an object).

wsfulton commented on Examples/Makefile.in:
```diff
@@ -904,13 +904,12 @@ NOLINK ?= false
+	(test -f swigp4.ml  || $(SWIG) -ocaml -co swigp4.ml 2>/dev/null) &&  \
+	(test -f swig.cmi   || $(OCC) -c swig.mli) &&  \
+	(test -f swig.cmo   || $(OCC) -c swig.ml) &&  \
+	(test -f swigp4.cmi || $(OCC) -I $(OCAMLP4WHERE) -pp "camlp4o pa_extend.cmo q_MLast.cmo" -c swigp4.ml)
```
Isn't this `test` redundant now for the test-suite?
For the examples it should always be run otherwise it may be out of date.

ZackerySpytz replied:
> Isn't this test redundant now for the test-suite?

No, it's not redundant: OCAMLCORE is always run before each test case, so the `test`s are needed to prevent unnecessary operations (and race conditions).

> For the examples it should always be run otherwise it may be out of date.

It seems that would only occur if someone was tinkering with the swig library files.
I had felt that one should run `make clean-ocaml-examples`  or `make clean` in the individual example dirs to remove the old/outdated files.
So, I'm not sure if a change should be made here to the proposed code.

wsfulton commented on Examples/test-suite/ocaml/Makefile.in:
```diff
@@ -85,9 +86,24 @@ include $(srcdir)/../common.mk
+swigp4.cmi:
+	env SWIG_LIB=$(SWIG_LIB_DIR) $(SWIGEXE) -ocaml -co swigp4.ml
+	$(OCC) -I $(OCAMLP4WHERE) -pp "camlp4o pa_extend.cmo q_MLast.cmo" -c swigp4.ml
+extra_obs: swig.cmi swig.cmo swigp4.cmi
```
Also I don't think make clean will remove the swig.mli, swig.mli, and swgip4.ml files and whatever else invoking $(OCC) on these files.

wsfulton replied:
Ignore that, ocaml_clean from Examples/Makefile.in is invoked and that looks like it cleans up.

ZackerySpytz commented:
> You sure about 4e4bb54 not being used?

Sorry, that was my fault -- I was looking at the wrong build log...
